compiler/noirc_frontend/src/hir_def/types/arithmetic.rs

```diff
@@ -63,72 +67,113 @@ impl Type {
     ///
     /// Otherwise also attempt try_simplify_partial_constants, sort_commutative,
     /// and other simplifications
-    fn canonicalize_helper(&self, found_checked_cast: bool, run_simplifications: bool) -> Type {
+    fn canonicalize_helper(
+        &self,
+        bindings: &TypeBindings,
+        found_checked_cast: bool,
+        run_simplifications: bool,
+    ) -> Type {
         match self {
             Type::InfixExpr(lhs, op, rhs, inversion) => {
                 let kind = lhs.infix_kind(rhs);
                 let dummy_location = Location::dummy();
-                // evaluate_to_field_element also calls canonicalize so if we just called
-                // `self.evaluate_to_field_element(..)` we'd get infinite recursion.
-                if let Ok(lhs_value) =
-                    lhs.evaluate_to_field_element_helper(&kind, dummy_location, run_simplifications)
-                {
-                    if let Ok(rhs_value) = rhs.evaluate_to_field_element_helper(
+
+                let evaluate = |typ: &Type| {
+                    typ.evaluate_to_field_element_helper(
                         &kind,
                         dummy_location,
+                        bindings,
                         run_simplifications,
-                    ) {
+                    )
+                };
+
+                // evaluate_to_field_element also calls canonicalize so if we just called
+                // `self.evaluate_to_field_element(..)` we'd get infinite recursion.
+                if let Ok(lhs_value) = evaluate(lhs) {
+                    if let Ok(rhs_value) = evaluate(rhs) {
                         if let Ok(result) = op.function(lhs_value, rhs_value, &kind, dummy_location)
                         {
                             return Type::Constant(result, kind);
                         }
                     }
                 }
 
-                let lhs = lhs.canonicalize_helper(found_checked_cast, run_simplifications);
-                let rhs = rhs.canonicalize_helper(found_checked_cast, run_simplifications);
+                // See if this is `X * 1` or `X / 1` in which case we can simplify it to `X`
+                if matches!(op, BinaryTypeOperator::Multiplication | BinaryTypeOperator::Division) {
+                    if let Ok(rhs_value) = evaluate(rhs) {
+                        if rhs_value.is_one() {
+                            return *lhs.clone();
+                        }
+                    }
+                }
+
+                // See if this is `X + 0` or `X - 0`, in which case we can simplify it to `X`
+                if matches!(op, BinaryTypeOperator::Addition | BinaryTypeOperator::Subtraction) {
+                    if let Ok(rhs_value) = evaluate(rhs) {
+                        if rhs_value.is_zero() {
+                            return *lhs.clone();
+                        }
+                    }
+                }
```

Comment:
I don't know if the two new rules above should happen here or somewhere else.

Reply:
Missed this part in the review - can you move these checks after the `let lhs = lhs.canonicalize_helper(...);` calls?

Currently if rhs is one or zero we're returning `lhs` but we're not canonicalizing it recursively in that case so the expression as a whole may not be in canonical form.